src/statemachine/coroutine.js

```diff
@@ -0,0 +1,189 @@
+pc.extend(pc, (function () {
+
+    var coroutines = [];
+
+    function noop() {
+    }
+
+    pc.ComponentSystem.on('update', function (dt) {
+        for (var i = coroutines.length - 1; i >= 0; i--) {
+            try {
+                coroutines[i]._step(dt);
+            }
+            catch (e) {
+                console.error(e);
+            }
+        }
+    });
+
+    /**
+     * @event
+     * @name pc.Coroutine#ended
+     * @description Fired when the coroutine is terminated for any reason
+     */
+
+    /**
+     * @callback
+     * @name pc.Coroutine~callback
+     * @description Called every time the coroutine executes
+     * @param {Number} dt The delta time since the last Update call
+     * @param {pc.Coroutine} coroutine The coroutine currently executing
+     * @returns {undefined|Boolean|pc.Coroutine~callback|Number} The return value designates what
+     *     will happen next.<br><br> Returning undefined, true or any string will cause the
+     *     Coroutine to run on the next frame. Returning 'false' will immediately terminate the
+     *     coroutine. Returning a number will delay the next execution for that many seconds.<br>
+     *     You may also return a different function which will be subsequently used to be the body
+     *     of the coroutine, when you do this coroutine.pop() can be called to return the previous
+     *     function to execution.
+     */
+
+    /**
+     * @class
+     * @name pc.Coroutine~parameters
+     * @description options for coroutine creation
+     * @property {Number} duration Duration of the coroutine
+     * @property {Object} bind An object to bind the enabled state to
+     * @property {Number} delay Delay before the coroutine executes
+     */
+
+    /**
+     * @class
+     * @name pc.Coroutine
+     * @description Basic support for coroutines that run for a fixed period, either dependent on
+     *     an object or independently
+     * @constructor Creates a new coroutine
+     * @param {pc.Coroutine~callback} fn The function to be run every frame
+     * @param {pc.Coroutine~parameters} options Optional object with an enabled property whose
+     *     value will control the execution of the coroutine
+     * @property {pc.Coroutine~callback} fn The currently running functionå
+     * @returns {pc.Coroutine} The new coroutine
+     */
+    function Coroutine(fn, options) {
+        options = options || {};
+        this._fn = fn || noop;
+        this._bind = options.bind;
+        this._destroyDelay = options.duration;
+        this._callDelay = options.delay || 0;
+        coroutines.push(this);
+        Object.defineProperties(this, {
```

Comment:
`.bind` is very inefficient and leads to a bit more memory used.
If you can use upper score `var self = this` - it is preferable option, then browsers can cache anonymous function, and improve its calling performance.
Generally `.bind` is pretty bad thing to use, and upper scope variable always preferable, unless it is very-very necessary.